### PR TITLE
Implement balance RPC methods that support Unified Addresses

### PR DESCRIPTION
**Pull request by str4d**

### CODE REVIEW

**Review by nuttycom — APPROVED** [review]
utACK with a potentially controversial opinion about the legacy account; I think this is good to merge as-is and then we can decide on that opinion after some good vigorous discussion.

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -3705,19 +3702,54 @@ UniValue z_getbalanceforaccount(const UniValue& params, bool fHelp)
     if (!fExperimentalOrchardWallet) {
         throw JSONRPCError(RPC_WALLET_ENCRYPTION_FAILED, "Error: the Orchard wallet experimental extensions are disabled.");
     }
-    int64_t account = params[0].get_int64();
+
+    int64_t accountInt = params[0].get_int64();
+    if (accountInt < 0 || accountInt >= ZCASH_LEGACY_ACCOUNT) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid account number, must be 0 <= account <= (2^31)-2.");
```
[the remark this answers is not in the document]
Comment (str4d) [reply src/wallet/rpcwallet.cpp:3750]: I disagree on this point, but I concur that we can expand the API later to support this if we decide it is the correct way to treat the legacy account number, whereas we can't remove support later if we added it now.

**Review by LarryRuane — APPROVED** [review]
ACK 756f4fc84057be486465e4f042d4b26becdbecc5
A couple of nonblocking suggestions, ran tests, looks good.

File: qa/rpc-tests/wallet_accounts.py [hunk 2]
```diff
@@ -78,10 +95,18 @@ def run_test(self):
         assert_equal(tx_details['outputs'][0]['type'], 'sapling')
         assert_equal(tx_details['outputs'][0]['address'], ua0)
 
+        # The new balance should not be visible with the default minconf, but should be
+        # visible with minconf=0.
         self.sync_all()
+        self.check_account_balance(0, {})
+        self.check_account_balance(0, {'sapling': 10}, 0)
```
[the remark this answers is not in the document]
Comment (str4d) [reply qa/rpc-tests/wallet_accounts.py]: This was intentional. The act of writing the first commit exposed the bug that the second commit fixes; I wanted to preserve this in the history.

This is a similar approach to "red-green" commit sequences where you first write the test that fails, and then write the code that makes it pass. More relatably, it's what I often do when updating test vectors: first commit the new test vectors so it's clear they fail, and then make the commit that causes them to pass.

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -3641,18 +3641,60 @@ UniValue z_getbalanceforaddress(const UniValue& params, bool fHelp)
...
+
     int minconf = 1;
     if (params.size() > 1) {
         minconf = params[1].get_int();
         if (minconf < 0) {
             throw JSONRPCError(RPC_INVALID_PARAMETER, "Minimum number of confirmations cannot be less than 0");
         }
     }
+
+    LOCK(pwalletMain->cs_wallet);
```
[the remark this answers is not in the document]
Comment (str4d) [reply src/wallet/rpcwallet.cpp:3660]: We cannot reduce its scope any further with the current `CWallet` interfaces. This `LOCK` is immediately before the first time we access `pwalletMain`.

**Pull request merged**